Stepper: Add Free Flow Scaffolding

edanzer opened the pull request:
### Proposed Changes

We are adding a simplified Free Flow for users to create a free site. It is based on stepper and mimics our recent tailored flows like newsletter and link in bio. The eventual goal is something close to proposed designs in this P2 about the Free Flow: p9Jlb4-5ST-p2. 

For this PR, the goal is NOT a ready-to-launch flow, but simply to get scaffolding in place so that we can define, work, and coordinate on discrete tasks. I created a [Tracking Task](https://github.com/Automattic/wp-calypso/issues/70491) with known follow ups. 

Specific changes: 
- Add a new "Free" flow
- Update the tailored flow "Intro" to allow starting a Free site
- Add a new "Free Setup" screen
- Add basic checklist items and translations to render Launchpad for Free Flow
- Updates to packages to set Launchpad screen settings for new Free sites

### Testing Instructions

**Question: This PR creates a new flow that is not ready for production. Should it be put behind a feature flag?**

Because this task is designed to setup scaffolding to clear the way for other tasks, not everything is working correctly. Just for clarity, I've noted many known issues below. All of these are addressed in the [Free Flow Tracking Task](https://github.com/Automattic/wp-calypso/issues/70491).

Because this flow is not fully working, the **real feedback needed is on whether what's included here provides a reasonable basis for others to work from, or whether there are other things that be included in this initial scaffolding.**

1) Check out this branch and run yarn and yarn start if needed. 

2) Go to this URL to start the process for setting up a site: 
http://calypso.localhost:3000/setup/free/intro

3) Go through the steps all the way to Launchpad.
   - You should get through all steps correctly. 
   - Known Issue: The Setup screen does not yet have a field for generating a domain name.
   - Known Issue: The Choose a Design screen is re-using the screen for Link in Bio. We will ultimately replace this with a new screen and different selection of themes.
   - Known Issue: You will see Domains and Plans screens. We will ultimately remove those.
  
4) On Launchpad, screen appearance and tasks should roughly match the launchpad screen shot in the P2 about the Free Flow: p9Jlb4-5ST-p2.
   - Known Issue: The link to Personalize Your Website is not active. That screen still needs to be created. 
   - Known Issue: You can click Write Your First Post. It will take you to the post editor. But if you go to My Home, which normally redirects to Launchpad, it will redirect to the "Goals" onboarding screen. You'll need to use the back button or manually enter the Launchpad url to return. Also, the Write Your First Post task will not be completed. 
   - Known Issue: The link to Edit Design does not currently go anywhere.
   - Known Issue: The Launch Your Site link will work, but you will end up redirected again to the "Goals" onboarding screen.
   
5) **UPDATE**: Adding extra testing step. Please consider testing newsletter and link in bio flows to ensure no regressive breakages. By definition, this PR has to touch files / methods / components used by those flows.

edanzer committed:
Add initial code for free stepper flow

edanzer committed:
Add launchpad tasks for free flow

edanzer committed:
Add launchpad strings for free flow

edanzer committed:
Set launchpad screen setting for free sites

jeyip commented on client/landing/stepper/declarative-flow/free.ts:
```diff
@@ -0,0 +1,140 @@
+		return [
+			'intro',
+			'linkInBioSetup',
+			'domains',
```
~`NB`: We don't want domains and plans for this flow~

Aaaand I just saw this mentioned in your PR description 👍

edanzer replied:
Yup, there's some complexity to removing them. I figured we'd handle each of those separately. I've added tasks for those to the Free Flow tracking issue.

jeyip commented:
> Question: This PR creates a new flow that is not ready for production. Should it be put behind a feature flag?

I'd suggest that we add a feature flag. It's not difficult to toggle flags, and the feature isn't ready to for a general audience.

jeyip commented:
To add to the list of issues:

- The site identity setup step has link in bio specific language ![Screen Shot 2022-11-29 at 3 30 32 PM copy](https://user-images.githubusercontent.com/5414230/204671442-19a9d55c-7d58-4bd4-b058-b8464f32f4df.jpg)

jeyip commented:
I can't seem to access the Launchpad

![2022-11-29 15 35 35](https://user-images.githubusercontent.com/5414230/204672114-5a713dfd-c168-4217-9c50-1f9d9e03b155.gif)

jeyip commented on client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/translations.tsx:
```diff
@@ -25,6 +30,12 @@ export function getLaunchpadTranslations( flow: string | null ): TranslatedLaunc
 				"All that's left is to add some links and launch your site."
 			);
 			break;
+		case FREE_FLOW:
```
You can add these to the translations specs

edanzer replied:
Added to specs!

jeyip commented on client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-helper.ts:
```diff
@@ -133,6 +143,31 @@ export function getEnhancedTasks(
 						},
 					};
 					break;
+				case 'site_launched':
```
I can't figure out what we changed here. Could you elaborate @edanzer ?

edanzer replied:
@jeyip We added a site_launched task to tasks.tsx and here. This task is nearly the same as link-in-bio-launched, but more more generic. 

As an alternative, we could also try to re-use link-in-bio-launched - we'd need to adjust that to be a general site launch task, which may have a bit of conditional handling within it for different flows that need a site launch task. Let me know if you think that's worth pursuing.

jeyip commented on client/landing/stepper/declarative-flow/internals/steps-repository/free-setup/index.tsx:
```diff
@@ -0,0 +1,111 @@
+	const { __ } = useI18n();
+	const site = useSite();
+
+	const formText = {
```
These placeholders don't seem to be respected see https://github.com/Automattic/wp-calypso/pull/70350#issuecomment-1331452659

edanzer replied:
Resolved. I was actually still referencing the link in bio setup form, rather than the new free flow setup form.

jeyip commented on client/landing/stepper/declarative-flow/free.ts:
```diff
@@ -0,0 +1,140 @@
+		const locale = useLocale();
+
+		// trigger guides on step movement, we don't care about failures or response
+		wpcom.req.post(
```
For those wondering why we need this post request "This calls a new endpoint in D87638-code that allows for reminding users about an uncompleted flow after a few days."

jeyip reviewed (commented):
The code and functionality looks fine, especially considering that most of it is boilerplate for adding a new flow.

I'll probably give it another once over in a day, because my main concern isn't bugs, but the accidental inclusion of extra logic that we don't need. @edanzer could you do me a favor and note any specific updates that are new, that haven't been included in previous flows, and that you think we should pay special attention to?

Addison-Stavlo reviewed (commented):
Thanks for starting this Erick!

Im curious how difficult (or simple) it would be to add the design picker step to this flow (replace the LIB patterns step with it) ? Do you have a sense of that at this point?

I think that could be beneficial to spike on briefly before we start digging into the dirt of removing plans/domains and incorporating the domain name generator. At least it might help us with determining a more accurate project ETA.

Addison-Stavlo commented:
> Known Issue: The Launch Your Site link will work, but you will end up redirected again to the "Goals" onboarding screen.

Thats funky! Do we have a sense of why it does this? or is that still tbd?

edanzer committed:
Add free-flow feature flag

edanzer commented:
> I'd suggest that we add a feature flag. It's not difficult to toggle flags, and the feature isn't ready for a general audience.

@jeyip - This has been added.

agrullon95 commented:
This a great start! We should add a feature flag and then merge so we can plan the next steps for the free onboarding flow.
The flow looks similar to the designs with the exception of the known issues you listed above.

The known issues related to redirecting back to "Goals" instead of "Home" can be fixed by updating the site's intent to 'free'.

edanzer committed:
Add free flow translations to specs

edanzer committed:
Ensure Free flow uses Free Setup

edanzer committed:
Fix missing styling for Free Flow

edanzer commented:
> Im curious how difficult (or simple) it would be to add the design picker step to this flow (replace the LIB patterns step with it) ? Do you have a sense of that at this point?

@Addison-Stavlo - I don't *think* it should be terribly difficult. As I think you know, the existing flows for setting up a site access the design picker step already, and you can probably pull it up for a site using this url: https://http://calypso.localhost:3000/setup/site-setup/designSetup?siteSlug=[SITESLUGHERE]. 

I have tried just inserting that step, but it doesn't quite work. So there is some extra info/context/adjustments needed for that step that I'm not sure on. If we like, I can continue to try to play with that to get that step inserted and working in place of the link-in-bio design picker. Otherwise, we can merge this as is, and create a follow up issue (which had been my plan).

edanzer commented:
To summarize, I've made the following recent changes in response to comments/feedback:
- Added a feature flag. Given its importance to allowing us to iterate on this feature, I'd appreciate others checking the feature flag and how I use it to confirm all is good. 
- Fixed issue with Setup page showing Link in Bio language. This was actually a bigger issue with PR still referencing Link in Bio setup page rather than the new Free Setup page. 
- Fixed some issues with missing styling. 
- Added specs for new Free translations. 
- Tested adding the general design picker in place of the link in bio design picker. This was not working. Rather than continue to implement and troubleshoot that in this PR (which is already big enough), I think we can tackle that as a follow up issue. 
- Rebased and resolved some merge conflicts. Rebasing also resolved some broken tests.

agrullon95 commented:
### Tests: 

- [ ] Feature flag `signup/free-flow` redirects the user to `/start/free` if disabled.
- [ ] Newsletter & Link-in-bio working without any regressive breakage.
- [ ] Navigating to `/home` redirects to LaunchPad


This is the only CSS issue we should fix before merging since it affects an entirely different page, besides that we should be good to merge this PR.


> If I negate the feature flag logic to disable it on the development environment, I get redirected to the `/start/free` page and the progress bar suddenly appears at the top of the page:
> 
> <img alt="image" width="960" src="https://user-images.githubusercontent.com/10482592/205740680-e94b6d7a-1305-406e-9495-d11598214217.png">

edanzer committed:
Remove progress bar on free domains page

edanzer commented:
> @edanzer - This is the only CSS issue we should fix before merging since it affects an entirely different page, besides that we should be good to merge this PR.

@agrullon95 - This has been fixed. Good catch on this. Took me a bit to figure out what was going on as I don't ever directly edit that screen (it's in the old start flow vs stepper flow we're typically working with now).

Addison-Stavlo commented on client/landing/stepper/declarative-flow/internals/steps-repository/processing-step/index.tsx:
```diff
@@ -117,7 +118,7 @@ const ProcessingStep: Step = function ( props ) {
 
 	// Currently we have the Domains and Plans only for link in bio
-	if ( flowName === LINK_IN_BIO_FLOW ) {
+	if ( flowName === LINK_IN_BIO_FLOW || FREE_FLOW ) {
```
I imagine this is intended to be something like:
```suggestion
	if ( flowName === LINK_IN_BIO_FLOW || flowName === FREE_FLOW ) {
```
? As what we have now seems like it will always be truthy.

Also noting throughout here I see a mix of `flowName === FREE_FLOW` and `isFreeFlow( flowName )`. Is there a reason we cant be consistent with one?

edanzer replied:
Good catch @Addison-Stavlo. This has been resolved. Now using isFreeFlow( flowName ) for the conditional.

edanzer replied:
@Addison-Stavlo - You've also got a good point on the `flowName === FREE_FLOW` vs `isFreeFlow()` conditionals. In order to minimize changes on a big PR so we can get this out, I'd probably prefer to leave those as is just for this PR. But I'll make a follow up for myself immediately after to go back through and update them all to isFreeFlow(). If we have that method, we should go ahead and use it. 

When I do that, I may clean up similar inconsistencies for `flowName === LINK_IN_BIO_FLOW` vs `isLinkInBioFlow()`. We're doing the same thing mixing those in the code.

Addison-Stavlo commented on client/landing/stepper/declarative-flow/internals/steps-repository/free-setup/styles.scss:
```diff
@@ -0,0 +1,16 @@
+		min-height: unset;
+	}
+
+	.free-setup .step-container__header {
```
Is this intended? it looks like we are looking for `.free-setup` inside `.free-setup` ?

Addison-Stavlo commented on client/landing/stepper/declarative-flow/internals/steps-repository/free-setup/styles.scss:
```diff
@@ -0,0 +1,16 @@
+		min-height: unset;
+	}
+
+	.free-setup .step-container__header {
```
Is this intended? It looks like we are looking for `.free-setup` inside `.free-setup`.

edanzer replied:
I had just mimicked the other setup forms. Oddly we do this for the link in bio and newsletter setup forms, too. And looking at the markup, it seems we actually do repeat .newsletter-setup / .link-in-bio-setup / .free-setup class on both the parent and on the step-container here. So both `.free-setup .step-container__header` and `.free-setup .free-setup .setup-container` are both valid. I'm hesitant to remove the extra class from markup in case it's used for other things, but clearly the extra selector is not needed here, so I've removed it.

edanzer committed:
Fix conditional bug

edanzer committed:
Remove unneeded css selector

Addison-Stavlo reviewed (approved):
This tests well on my end - with and without the feature flag as expected.  Looks safe to go after the recent changes.

agrullon95 commented:
The free flow is working well on my end.

I retested the following:
- [ ] Feature flag signup/free-flow redirects the user to /start/free if disabled.
- [ ] Newsletter & Link-in-bio working without any regressive breakage.
- [ ] Navigating to /home redirects to LaunchPad

I found CSS issues on the stepper domains page, but they seem to be unrelated to this PR:

<img width="793" alt="Screen Shot 2022-12-06 at 12 31 04 PM" src="https://user-images.githubusercontent.com/10482592/205987284-e93949f1-b6c2-4764-a5a9-90ae466368d6.png">

> `Show more results` button should be centered


<img width="747" alt="Screen Shot 2022-12-06 at 12 29 46 PM" src="https://user-images.githubusercontent.com/10482592/205987336-ed77c49b-3aeb-4471-96dd-8445bad26c1b.png">

> `Select` button should right aligned

jeyip commented:
I did one more quick walk-through:

## Testing
### Requirements
- [x] Site created properly ( with correct site title )
- [x] Selected site design is installed
- [x] Launchpad screen is "generally" styled and laid out correctly 

### Browsers
- [x] Chrome

We can follow-up and tackle a lot of these issues in follow-ups since this feature is gated behind a flag. 🚢

edanzer committed:
Fix styling issue on Free launchpad screen

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7661196

Hi @edanzer, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include this string: `Upgrade to a Premium plan to publish your color changes and unlock tons of other features.`

Thank you in advance!

escapemanuele commented:
Hey @Automattic/cylon 👋 
Just a heads up that I updated your `free` flow in our [latest PR ](https://github.com/Automattic/wp-calypso/pull/70478),  to load steps individually and separately for each flow. I hope we're shipping this soon.
If you update your flow, feel free to reach out to me so I can keep it up to date when rebasing. I also will tell you when we ship our PR :)

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7661196).
